core/types/common.go

```diff
@@ -16,10 +16,28 @@
 
 package types
 
-import "github.com/ethereum/go-ethereum/core/vm"
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
 
 type BlockProcessor interface {
 	Process(*Block) (vm.Logs, Receipts, error)
 	ValidateHeader(*Header, bool, bool) error
 	ValidateHeaderWithParent(*Header, *Header, bool, bool) error
 }
+
+type Validator interface {
+	ValidateBlock(block *Block) error
+	ValidateHeader(header, parent *Header, checkPow bool) error
+	ValidateState(block, parent *Block, state *state.StateDB, receipts Receipts, usedGas *big.Int) (ethdb.Batch, error)
+}
+
+type Processor interface {
+	Process(block *Block, statedb *state.StateDB) (Receipts, vm.Logs, *big.Int, error)
+}
+
+type BlockValidatorFn func(*Block) (*state.StateDB, ethdb.Batch, Receipts, error)
```

Comment:
I think these interfaces should be in package core.
Having them in types creates these really weird dependencies on state and vm.